scanner/src/funTest/kotlin/ScannerFunTest.kt

```diff
@@ -0,0 +1,217 @@
+
+/**
+ * Regular expression to match temporary download paths in a scan result. As these paths change on each test run,
+ * they have to be dealt with when comparing results.
+ */
+private val DOWNLOAD_PATH_REGEX = Regex("-download(\\d+)/")
+
+/**
+ * Regular expression to match the *variables* element in a scan result. Variables depend on the current
+ * environment; so they have to be deleted when comparing results.
+ */
+private val VARIABLES_REGEX = Regex("variables:.*?tool_versions:", RegexOption.DOT_MATCHES_ALL)
+
+/** Some labels to assign to the test result. */
+private val labels = mapOf(
+    "test" to "true",
+    "scan" to "full"
+)
+
+/** Stores a result object with the expected scanner result to compare to in tests. */
+private var expectedScannerResult = constructExpectedResult()
+
+/**
+ * Patch the given [scan result][result] by replacing variables part, so that it can be compared.
+ */
+private fun patchResult(result: String): String =
+    stripTempPaths(stripVariables(result))
+
+/**
+ * Replace temporary path names in the given [result].
+ */
+private fun stripTempPaths(result: String): String =
+    result.replace(DOWNLOAD_PATH_REGEX, "/")
+
+/**
+ * Remove the section with environment variables from the given [result].
+ */
+private fun stripVariables(result: String): String =
+    result.replace(VARIABLES_REGEX, "tool_versions:")
+
+/**
+ * Read the file with the expected scanner result, replace variable parts, and convert it to an [OrtResult] object.
+ */
+private fun constructExpectedResult(): OrtResult {
+    val expectedResultYaml = patchResult(
+        patchExpectedResult(
+            assetsDir.resolve("test-scanner-expected-output-for-analyzer-result.yml"),
+            revision = "HEAD"
+        )
+    )
+
+    return deserializeResult(expectedResultYaml)
+}
+
+/**
+ * Compares the [result] specified against the expected scanner result. Note that comparison needs to be done on
+ * object level, as the raw YAML representation may differ. The [result] passed in needs to be patched first to
+ * deal with variable parts. Also, access statistics cannot be compared directly.
+ */
+private fun checkResult(result: OrtResult) {
+    val patchedYaml = patchResult(
+        patchActualResult(
+            yamlMapper.writeValueAsString(result),
+            patchDownloadTime = true,
+            patchStartAndEndTime = true
+        )
+    )
+    val patchedResult = deserializeResult(patchedYaml)
+    val statistics = patchedResult.statistics()
+    val expectedStatistics = expectedScannerResult.statistics()
+    statistics.numHits.get() shouldBe expectedStatistics.numHits.get()
+    statistics.numReads.get() shouldBe expectedStatistics.numReads.get()
+    // AtomicInteger does not implement equals(); so copy objects to make comparison succeed.
+    statistics.numHits = expectedStatistics.numHits
+    statistics.numReads = expectedStatistics.numReads
+
+    patchedResult shouldBe expectedScannerResult
```

Comment:
The statistics object is problematic because of the AtomicIntegers it contains. AtomicInteger does not implement equals(). Therefore, the test checks these values separately and then overwrites them, to make the comparison of the whole result object succeed.